File: apps/android/app/src/main/java/com/suzent/mobile/StreamRecovery.kt
package com.suzent.mobile

import java.io.IOException
import org.json.JSONObject

class StreamRecovery {
    var runId: String? = null
        private set
    var sequence = 0L
        private set
    var ended = false
        private set
    var persisted = false
        private set
    var superseded = false
        private set

    fun request(chatId: String): JSONObject = JSONObject().put("chat_id", chatId)
        .put("wait_ms", 1000).put("protocol", 1).apply {
            runId?.let { put("run_id", it); put("after_seq", sequence) }
        }

    fun consume(frame: JSONObject): List<JSONObject> {
        val run = frame.optString("run_id")
        val seq = frame.optLong("seq", -1)
        if (run.isEmpty() || seq < 0) throw IOException("Invalid recovery frame")
        return when (frame.optString("type")) {
            "STREAM_SNAPSHOT" -> {
                val events = frame.optJSONArray("events") ?: throw IOException("Invalid snapshot")
                val decoded = (0 until events.length()).map { events.getJSONObject(it) }
                runId = run; sequence = seq
                listOf(JSONObject().put("type", "STREAM_RESET")) + decoded
            }
            "STREAM_EVENT" -> {
                if (run != runId) throw IOException("Unexpected stream run")
                if (seq <= sequence) emptyList() else {
                    if (seq != sequence + 1) throw IOException("Missing stream event")
                    val event = frame.getJSONObject("event")
                    sequence = seq
                    listOf(event)
                }
            }
            "STREAM_END" -> {
                if (run != runId || seq != sequence) throw IOException("Invalid stream completion")
                ended = true
                persisted = frame.optBoolean("persisted")
                superseded = frame.optBoolean("superseded")
                emptyList()
            }
            else -> throw IOException("Unsupported recovery protocol")
        }
    }
}

File: packages/SuzentCore/Sources/SuzentCore/StreamRecovery.swift
import Foundation

struct StreamFrame: Decodable {
    let type: String
    let run_id: String
    let seq: Int
    let events: [StreamEvent]?
    let event: StreamEvent?
    let persisted: Bool?
    let superseded: Bool?
}

struct StreamRecovery {
    private(set) var runID: String?
    private(set) var sequence = 0
    private(set) var ended = false
    private(set) var persisted = false
    private(set) var superseded = false

    func request(chatID: String) throws -> Data {
        var body: [String: Any] = ["chat_id": chatID, "wait_ms": 1000, "protocol": 1]
        if let runID { body["run_id"] = runID; body["after_seq"] = sequence }
        return try JSONSerialization.data(withJSONObject: body)
    }

    mutating func consume(_ frame: StreamFrame) throws -> [StreamEvent] {
        guard !frame.run_id.isEmpty, frame.seq >= 0 else { throw ClientError.invalidResponse }
        switch frame.type {
        case "STREAM_SNAPSHOT":
            guard let events = frame.events else { throw ClientError.invalidResponse }
            runID = frame.run_id; sequence = frame.seq
            return [StreamEvent(type: "STREAM_RESET", delta: nil, message: nil)] + events
        case "STREAM_EVENT":
            guard frame.run_id == runID else { throw ClientError.invalidResponse }
            if frame.seq <= sequence { return [] }
            guard frame.seq == sequence + 1, let event = frame.event else { throw ClientError.invalidResponse }
            sequence = frame.seq
            return [event]
        case "STREAM_END":
            guard frame.run_id == runID, frame.seq == sequence else { throw ClientError.invalidResponse }
            ended = true; persisted = frame.persisted == true; superseded = frame.superseded == true
            return []
        default: throw ClientError.invalidResponse
        }
    }
}
